docs/src/appendix/mast.md

```diff
@@ -0,0 +1,526 @@
+    /// field of `MastNodeInfo`
+    External,
+}
+```
+
+The only thing to note here is that we want the `MastNodeInfo` type to be fixed-width, to enable indexing. We also want to keep the encoding compact, but efficient. As a result, we are making a slight tradeoff with the encoding of this type to make traversal fast, by inlining references to other nodes. This means that there are padding bytes introduced for node types which do not have such references, and which cannot take advantage of the padding to store a subset of their data. This is estimated to be a worthwhile tradeoff, but isn't based on any benchmarks, just gut feeling.
+
+##### Basic Blocks
+
+Unlike the other node types, `BasicBlockNode` has a much more complex representation. It must encode the operations contained in the block, as well as decorators for those operations. Both operations and decorators may have associated data of their own. Additionally, operations of a block are batched and grouped to facilitate computation of their Merkle digests.
+
+We do not represent batching/grouping in the binary format, at least in this version. Instead, we specify that operations decoded from the binary format will be batched and grouped using fixed sizes. Specifically, a batch size of 8, and a group size of 9. This could alternatively be handled by storing these sizes in the unused padding bytes in `MastNodeType` which are available to the `Block` variant. For this draft though, I'm making this a property of the specific version of the binary format, rather than per-block configurable.
+
+```rust
+/// Represents a variable-width index into the `strings` table
+///
+/// If the strings table is small, indexes into it are only a
+/// single byte, but if the table grows too large, only those
+/// later strings need pay the price of the larger index.
+pub type StringIndex = VInt64;
+
+/// An entry in the `strings` table of `EncodedMastForest`
+///
+/// We could also use null-terminated strings, but this
+/// allows us to be more precise, and permit any content
+/// in the strings, even NUL bytes.
+#[repr(C)]
+pub struct StringRef {
+    offset: DataOffset,
+    // Maximum allowed length of a string is 2^32
+    len: u32,
+}
+
+/// Unlike most other type definitions here, this one
+/// is not encoded using the usual Rust layout rules.
+///
+/// Specifically, we treat this as a variable-width type,
+/// without the padding that would normally be introduced
+/// to make smaller variants as large as the largest variant.
+///
+/// So each encoded operation starts with the tag byte, which
+/// indicates which variant we're going to be parsing. We then
+/// parse the data for that variant, and immediately look for
+/// the next tag byte (if there are more ops to parse).
+///
+/// It is strictly for purposes of readability.
+///
+/// NOTE: Decorators are always associated with the next non-decorator
+/// operation in the basic block being decoded.
+#[repr(u8)]
+pub enum EncodedOperation {
```

Comment:
Maybe the comment explains exactly this, but I think we don't need a special enum to differentiate different types of operations and decorators. Specifically, we just need 1 byte to encode both the tag and the opcode.

This is possible because operations require only 7 bits to encode. So, we can say that if the first bit of a byte is 0, the remaining 7 bits encode an opcode; if the first bit is 1, the remaining 7 bits encode tag of a decorator.

This way, vast majority of `EncodedOperation`'s would require 1 byte to encode (rather than 2 bytes).